src/Process/MemoryLimitSetter.php

```diff
@@ -0,0 +1,45 @@
+<?php
+
+namespace Meteor\Process;
+
+/**
+ * Class MemoryLimitSetter
+ *
+ * @author Jadu Ltd.
+ */
+class MemoryLimitSetter
+{
+    const RE_PHP_SCRIPT = '/^\s*?(php)/';
+    const RE_MEMORY_LIMIT = '/([-]{1,2}(?:define|d)\s*(memory_limit))/';
+
+    /**
+     * @param string $command
+     * @return boolean
+     */
+    public function isPHPScript($command)
+    {
+        return preg_match(self::RE_PHP_SCRIPT, $command);
+    }
+
+    /**
+     * @param string $command
+     * @return boolean
+     */
+    public function hasMemoryLimit($command)
+    {
+        return preg_match(self::RE_MEMORY_LIMIT, $command);
+    }
+
+    /**
+     * @param string $command
+     * @return string
+     */
+    public function setMemoryLimit($command)
```

Comment:
This should probably be instantiated with the `$command` and have a factory for dependency injection?

Reply:
Alternatives could be to directly wrap the `Process` class as a `PhpProcess` or to set the methods to static

Reply:
Do you mind explaining why those approaches would be preferable here Jack? I agree that we could make those methods static. I didn't want to introduce any inheritance or to refactor too much of the existing process behaviour.

Reply:
With this said, I can see the point here,  couldew the declarative nature but have those methods live on the process `$process->isPHPScript()` ect... perhaps I will look at that actually, I wasn't planning on making too many changes to existing stuff though – will think about this

Reply:
Also on having a factory method, I've opted to follow the meteor convention (https://github.com/jadu/meteor/pull/120/files#diff-6eadfc058a6a2e608b71fc9212532dc6R70) which effectively acts as a factory and dependency injection, apologies if that's not what you meant

Reply:
Not sure how best to explain this. I tried to find an explanation of what I'm trying to say/pattern I'm suggesting but couldn't really find anything.

Basically two things here. First is that your `MemoryLimitSetter` is specific to PHP `Process`s only. So the name somewhat doesn't make sense. Also, your class name is an action rather than something that can invoke an action - if that makes sense?

The second thing that I tried to explain is the pattern. This is kind of the best explanation I could find: https://stackoverflow.com/a/4367410/1600814 - Java code tends to have a lot of `Util` classes which is similar to how this class is currently constructed

Essentially you currently have this:
```php
$class = new Class();
if ($class->some($name)) {
    $name = $class->thing($name);
}
$process = new Process($name);
```

So the main issue with this is that the construction of the class is pointless - other than enabling unit testing of the dependency injection (which could be argued pointless unit testing also, or probably more likely unit testing the wrong thing in the wrong place).

So something better would be to use static methods:
```php
if (Class::some($name)) {
    $name = Class::thing($name);
}
$process = new Process($name);
```

Or you could pass the duplicated arguments into the constructor which would then make the object less "util" and specific to the variable its self. This helps reduce duplication.

```php
$class = new Class($name);
if ($class->some()) {
    $name = $class->thing();
}
$process = new Process($name);
```

What I meant about a factory is that once you have this form, you'd need a factory class to inject for proper DI and for unit testing.

```
$class = $classFactory->create($name);
```

But as we already know, the Class object (MemoryLimitSetter) is specific to the `Process` and even more specific, only to PHP processes. So you could quite reasonably extend the `Process` class as a `PhpProcess` class. This to me sounds like good inheritance as it's the same thing but narrowed down. You'd then want to decide to apply the memory limit likely at the point of `->run()` so that it can be enabled/disabled in the class. Then all you need to do is make the process factory aware of the types and create the correct one.

```
if (PhpProccess::isPhpCommand($command) {
   return new PhpProcess($command);
}

return new Process($command);
```

```php
$factory->create($comamnd);
// Enable by default
$process->disableMemoryLimit();
$process->enableMemoryLimit();
$process->run();
```

```php
class PhpProccess extends Process {
    public function run() {
        if ($this->limitMemory) {
            $this->alterCommand();
        }

        parent::run();
    }
}
```

Hopefully, any of this waffle makes sense 😂

Reply:
Cheers @jackbentley I really appreciate you taking to time to explain that in such detail. I agree in principle that we shouldn't instantiate that class unnecessarily, so I'll make those methods static.